internal/constraints/constraints.go

```diff
@@ -72,62 +42,36 @@ func NewConditional(a *authentication.Auth) *Conditional {
 	return c
 }
 
-type projectable interface {
-	Owner() string
-	Name() string
-	NamespaceString() string
-	CommitID() string
-	BranchName() string
-	Path() string
-	URL() string
-}
+func NewPrimeConditional(structure interface{}) *Conditional {
+	c := NewConditional()
 
-func NewPrimeConditional(auth *authentication.Auth, pj projectable, subshellName string) *Conditional {
-	var (
-		pjOwner     string
-		pjName      string
-		pjNamespace string
-		pjURL       string
-		pjCommit    string
-		pjBranch    string
-		pjPath      string
-	)
-	if !p.IsNil(pj) {
-		pjOwner = pj.Owner()
-		pjName = pj.Name()
-		pjNamespace = pj.NamespaceString()
-		pjURL = pj.URL()
-		pjCommit = pj.CommitID()
-		pjBranch = pj.BranchName()
-		pjPath = pj.Path()
-		if pjPath != "" {
-			pjPath = filepath.Dir(pjPath)
-		}
+	v := reflect.ValueOf(structure)
+	// deref if needed
+	if v.Kind() == reflect.Ptr {
+		v = v.Elem()
 	}
 
-	c := NewConditional(auth)
-	c.RegisterParam("Project", map[string]string{
-		"Owner":     pjOwner,
-		"Name":      pjName,
-		"Namespace": pjNamespace,
-		"Url":       pjURL,
-		"Commit":    pjCommit,
-		"Branch":    pjBranch,
-		"Path":      pjPath,
-
-		// Legacy
-		"NamespacePrefix": pjNamespace,
-	})
-	osVersion, err := sysinfo.OSVersion()
-	if err != nil {
-		multilog.Error("Could not detect OSVersion: %v", err)
+	fields := reflect.VisibleFields(v.Type())
+
+	// Work at depth 1: Vars.[Struct].Struct.Simple
+	for _, f := range fields {
+		d1Val := v.FieldByIndex(f.Index)
+		if d1Val.Kind() == reflect.Ptr {
+			d1Val = d1Val.Elem()
+		}
+
+		// Only nodes at depth 1 need to be registered since the generic type
+		// handling within the templating package will do the rest. If function
+		// registration is needed at greater depths, this will need to be
+		// reworked (and may not be possible without expansive refactoring).
+		switch d1Val.Type().Kind() {
+		case reflect.Func:
+			c.RegisterFunc(f.Name, d1Val.Interface())
+
+		default:
+			c.RegisterParam(f.Name, d1Val.Interface())
+		}
 	}
```

Comment:
If there is one part of this PR that needs thorough commenting it's this one.. Please revisit this code such that someone can read through it and understand what's happening without hitting the docs.